google/cloud/spanner_v1/streamed.py

```diff
@@ -143,7 +160,13 @@ def __iter__(self):
                     return
                 iter_rows, self._rows[:] = self._rows[:], ()
             while iter_rows:
-                yield iter_rows.pop(0)
+                row = iter_rows.pop(0)
+                if self._results_checksum is not None:
+                    self._results_checksum.consume_result(row)
+                    _compare_checksums(
+                        self._original_results_checksum, self._results_checksum
+                    )
+                yield row
```

Comment:
The java version actually consumes the results up to the same point that the user did before retrying before doing the comparison: https://github.com/googleapis/java-spanner/blob/059ef1ef1f03e80f4ff2705b45a62c553bb4e83d/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ChecksumResultSet.java#L145-L153. Here we actually yield the rows used in the comparison. How come this doesn't result in the user getting the same rows twice after an internal retry?

It's also a bit confusing to see this do the checksum comparison with each row. I know `_compare_checksums` passes if it sees fewer retried rows than original, but that's not what I'd expect it to do without reading it.

Reply:
@olavloite, @skuruppu mentioned you might be able to review and compare this to the java client behavior.

Reply:
> How come this doesn't result in the user getting the same rows twice after an internal retry?

Seems to me a user will not use the rows read within a transaction in operations outside the transaction until the transaction is finished. It probably sounds hard, so consider this case:

User starts a transaction, reads several rows, then calls his own (not Spanner) API which is using these rows to write some data somewhere. Then he calls more functions in the same transaction and commits it. It's incorrect way of use - one should not do anything with data which the transaction is using at that moment, as transaction can fail and all the data/changes will be rollbacked, but it'll not rollback the records made in the user API, so the user will get the data desynchronization.

With this, the data that was read within transaction are considered as a temporary inexact-until-transaction-commit data, and if we're retrying a transaction, all that data should be re-read and recalculated, that's how I see it.

Reply:
> It's also a bit confusing to see this do the checksum comparison with each row. I know _compare_checksums passes if it sees fewer retried rows than original, but that's not what I'd expect it to do without reading it.

Yeah, that's something I'm not 100% like too, but retry logic looks too apportioned across classes and functions to me, so I've decided to make many things deep under the hood. Thus, in the transaction class we're not thinking about results number of a result set of one of the operations of this transaction, we're just calling `==` and it takes care about such a things deeper in the result set itself.

Reply:
I have absolutely no Python experience, so bear with me if I misread or misunderstood (parts of) the code here, but if I understand it correctly the suggested implementation for Python here does the following (during the initial transaction):

1. The `__iter__` method will normally (i.e. in the original transaction) iterate over the `PartialResultSet`s that are received from Spanner, and for each of these iterate over the rows in that `PartialResultSet`.
1. If the `ResultSet` has a checksum, it will update the running checksum with the contents of the current row.
1. It then checks whether the current checksum is equal to the checksum of the original transaction. If this is the original transaction and not a retry, the value of `_original_results_checksum` will be `None` and the comparison will always succeed.
1. It then returns the row.

If a transaction is retried, the following happens:

1. The `__iter__` method will iterate over the `PartialResultSet`s that are received from Spanner for the retried query, and for each of these iterate over the rows in that `PartialResultSet`.
1. If the `ResultSet` has a checksum, it will update the running checksum with the contents of the current row.
1. It then checks whether the current checksum is equal to the checksum of the original transaction. As far as I can see, this would always fail if the result set contained more than 1 row, as it is a running checksum.

Assume that a query returns the following rows during the original transaction:

| Key | Value |
| - | - |
| 1 | One |
| 2 | Two |

During the original transaction the entire result set was consumed so the running checksum was calculated based on all the rows:

* After row 1 the checksum is `C1`
* After row 2 the checksum is `C2`
So the original checksum is now `C2`.

During a retry the query will be retried and the checksum will be calculated again (I assume that the results during the retry are equal to the original query):
* After row 1 the checksum is `C1`. That is compared to the original checksum that is `C2` and the comparison will fail.

The checksum should therefore not be compared for each row, but after consuming the same number of rows during the retry as during original query.

(But I get a feeling that there is a more fundamental misunderstanding about this retry logic. I'll elaborate on that in a separate comment.)

Reply:
Hi, @olavloite

> It then checks whether the current checksum is equal to the checksum of the original transaction. As far as I can see, this would always fail if the result set contained more than 1 row, as it is a running checksum.

Yeah, that's the thing @c24t mentioned as something that we expect. In fact I'm redefining `==` operator for `ResultsChecksum` object, so that it compares checksums in this way:
```python
if original is not None:
        if retried != original:
            if not retried < original:  # retried has less results
```
At the first line it checks if we're in retried transaction - if not, than we have nothing to compare.

The second checks if checksums are equal, considering their length (number of consumed results).

And the third line checks if the retried transaction camed to the original transaction failure point.

So, if the given checksums have different length, it'll be considered like there are no problems so far, 'cause a) retried transaction is not at the point, where the original failed, so it's too soon to compare checksums; b) retried transaction went farther than the original transaction = the checksums were compared = they were equal at the original transaction failure point).

Looking at this, I think there can be a problem when retried transaction went farther than the original. I'll double check

Reply:
@c24t, hm-m, I think the logic here can be little bit simpler, if I'll use a `__len__` magic method.